internal/controller/cluster_upgrade.go

```diff
@@ -697,39 +697,50 @@ func (r *ClusterReconciler) upgradeInstanceManager(
 		}
 		operatorHash := targetManager.GetHash()
 
-		if instanceManagerHash != "" && instanceManagerHash != operatorHash && !instanceManagerIsUpgrading {
-			// We need to upgrade this Pod
-			contextLogger.Info("Upgrading instance manager",
-				"pod", postgresqlStatus.Pod.Name,
-				"oldVersion", postgresqlStatus.ExecutableHash)
-
-			if cluster.Status.Phase != apiv1.PhaseOnlineUpgrading {
-				err := r.RegisterPhase(ctx, cluster, apiv1.PhaseOnlineUpgrading, "")
-				if err != nil {
-					return err
-				}
-			}
+		if instanceManagerIsUpgrading || instanceManagerHash == "" || instanceManagerHash == operatorHash {
+			message := fmt.Sprintf("Instance manager will skip upgrade on %s (upgrading: %t) "+
+				"(operator hash: %s — instance manager hash: %s)",
+				postgresqlStatus.Pod.Name,
+				instanceManagerIsUpgrading,
+				operatorHash[:6],
+				instanceManagerHash[:6])
+			contextLogger.Trace(message)
+			continue
+		}
 
-			err = r.InstanceClient.UpgradeInstanceManager(ctx, postgresqlStatus.Pod, targetManager)
+		// We need to upgrade this Pod
+		contextLogger.Info("Upgrading instance manager",
+			"pod", postgresqlStatus.Pod.Name,
+			"oldHash", instanceManagerHash,
+			"newHash", operatorHash)
+
+		if cluster.Status.Phase != apiv1.PhaseOnlineUpgrading {
+			err := r.RegisterPhase(ctx, cluster, apiv1.PhaseOnlineUpgrading, "")
 			if err != nil {
-				enrichedError := fmt.Errorf("while upgrading instance manager on %s (hash: %s): %w",
-					postgresqlStatus.Pod.Name,
-					operatorHash[:6],
-					err)
-
-				r.Recorder.Event(cluster, "Warning", "InstanceManagerUpgradeFailed",
-					fmt.Sprintf("Error %s", enrichedError))
-				return enrichedError
+				return err
 			}
+		}
 
-			message := fmt.Sprintf("Instance manager has been upgraded on %s (hash: %s — previous hash: %s)",
+		err = r.InstanceClient.UpgradeInstanceManager(ctx, postgresqlStatus.Pod, targetManager)
+		if err != nil {
+			enrichedError := fmt.Errorf("while upgrading instance manager on %s (hash: %s): %w",
 				postgresqlStatus.Pod.Name,
 				operatorHash[:6],
-				instanceManagerHash[:6])
+				err)
 
-			r.Recorder.Event(cluster, "Normal", "InstanceManagerUpgraded", message)
-			contextLogger.Info(message)
+			r.Recorder.Event(cluster, "Warning", "InstanceManagerUpgradeFailed",
+				fmt.Sprintf("Error %s", enrichedError))
+			return enrichedError
 		}
+
+		message := fmt.Sprintf("Instance manager has been upgraded on %s "+
+			"(oldHash: %s — newHash: %s)",
+			postgresqlStatus.Pod.Name,
+			instanceManagerHash[:6],
+			operatorHash[:6])
+
+		r.Recorder.Event(cluster, "Normal", "InstanceManagerUpgraded", message)
+		contextLogger.Info(message)
```

Comment:
@mnencia and this one is Info, I think that the one indicating skip should be Info for sure, from a user perspective